pandas/io/html.py

```diff
@@ -342,44 +345,149 @@ def _build_doc(self):
 
-    def _parse_raw_thead(self, table):
-        thead = self._parse_thead(table)
+    def _parse_raw_thead_tbody_tfoot(self, table_html):
+        """Given a table, return parsed header, body, and foot.
+           Header and body are lists-of-lists. Top level list is a list of
+           rows. Each row is a list of parsed elements.
+
+           Logic: Use <thead>, <tbody>, <tfoot> elements to identify
+                  header, body, and footer, otherwise:
+                  - Put all rows into body
+                  - Move rows from top of body to header only if
+                    all elements inside row are <th>
+                  - Move rows from bottom of body to footer only if
+                    all elements inside row are <th>
+
+        Parameters
+        ----------
+        table_html : a single HTML table element.
+
+        Returns
+        -------
+        tuple of (header, body, footer)
+        header : list of rows, each of which is a list of parsed header
+        elements
+        body : list of rows, each of which is a list of parsed body elements
+        footer : list of rows, each of which is a list of parsed footer
+        elements
+        """
+        header_rows = []
+        body_rows = []
+        footer_rows = []
+        # first, are there thead and tbody elements in the table?
+        if (self._contains_tag(table_html, 'thead') and
+                self._contains_tag(table_html, 'tbody')):
+            header_rows = self._extract_tr(self._extract_thead(table_html)[0])
+            body_rows = self._extract_tr(self._extract_tbody(table_html)[0])
+            if self._contains_tag(table_html, 'tfoot'):
+                footer_rows = self._extract_tr(
+                    self._extract_tfoot(table_html)[0])
+        else:
+            # otherwise we need to split the body into header/body/foot
+            body_rows = self._extract_tr(table_html)
+            if body_rows == []:
+                # empty table, just return nothing
+                return [], [], []
+            # splitting criterion: if all tags within a row are th, it's part
+            # of the header/footer
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[0])):
+                # this row should be a header row, move it from body to header
+                header_rows.append(body_rows.pop(0))
+            while all(self._equals_tag(t, 'th') for t in
+                      self._extract_td(body_rows[-1])):
+                # this row should be a footer row, move it from body to footer
+                footer_rows.insert(0, body_rows.pop())
+        header = self._expand_colspan_rowspan(header_rows, fill_rowspan=False)
+        body = self._expand_colspan_rowspan(body_rows, fill_rowspan=True)
+        footer = self._expand_colspan_rowspan(footer_rows, fill_rowspan=False)
+        # the below line is lifted from _parse_raw_tfoot. Not sure what it
+        # does.
```

Comment:
Add newline above your comment.  Rewrite so that "does" is not alone in this line.